### PR TITLE
fontconfig_210: remove package

### LINKED ISSUE #88289: Vulnerability roundup 84: fontconfig-2.10.2: 1 advisory
[search](https://search.nix.gsc.io/?q=fontconfig&i=fosho&repos=NixOS-nixpkgs), [files](https://github.com/NixOS/nixpkgs/search?utf8=%E2%9C%93&q=fontconfig+in%3Apath&type=Code)

* [ ] [CVE-2016-5384](https://nvd.nist.gov/vuln/detail/CVE-2016-5384) CVSSv3=7.8 (nixos-20.03)

Scanned versions: nixos-20.03: 82b5f87fcc7. May contain false positives.

### PR DESCRIPTION
**Pull request by flokli**
This removes the `fontconfig_210` package, as well as the only remaining usage, the rendering of 2.10.x config and cache.

I switched my system to this, and didn't spot any breakages.

###### Motivation for this change
Closes https://github.com/NixOS/nixpkgs/issues/88289

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by flokli** [comment]
There's a lot more of (now possibly unused) code - there's `make-fonts-conf.xsl` used to post-massage fontconfigs `/etc/fonts/font.conf` file to include `fontDirectories` - and to load files from the versioned subdir.

This is also exposed as `makeFontsConf` (which seems to be used in various places)

For now I decide to keep this, to not change too much at once, but I assume we might want to remove the configVersion parts in there, too.

**Review by jtojnar — APPROVED** [review]
This looks great but I would like to hold off on merging this until we have dealt with the possible fallback from https://github.com/NixOS/nixpkgs/pull/73795

**Comment by jtojnar** [comment]
> There's a lot more of (now possibly unused) code - there's `make-fonts-conf.xsl` used to post-massage fontconfigs `/etc/fonts/font.conf` file to include `fontDirectories` - and to load files from the versioned subdir.
> 
> This is also exposed as `makeFontsConf` (which seems to be used in various places)
> 
> For now I decide to keep this, to not change too much at once, but I assume we might want to remove the configVersion parts in there, too.

Yeah, the XSL stylesheet does two things at once. The reason why it is used in the fontconfig derivation is precisely to introduce the versioned `include` so we cannot remove the `fontconfigConfigVersion` variable or fontconfig would no longer load its global configuration.

**Comment by flokli** [comment]
@jtojnar has there been any breakage reported already, or do you mean you just want to await one staging cycle?

**Comment by flokli** [comment]
> Yeah, the XSL stylesheet does two things at once. The reason why it is used in the fontconfig derivation is precisely to introduce the versioned include so we cannot remove the fontconfigConfigVersion variable or fontconfig would no longer load its global configuration.

I'd propose removing the `fontconfigConfigVersion` parts from the xslt stylesheet, if the support fontconfig was the only (remaining?) consumer of it. (But in a followup PR after this one)

**Comment by jtojnar** [comment]
> @jtojnar has there been any breakage reported already, or do you mean you just want to await one staging cycle?

Yeah, I would like to wait one staging cycle, or maybe slightly more since the channels lag a bit behind.

> I'd propose removing the `fontconfigConfigVersion` parts from the xslt stylesheet, if the support fontconfig was the only (remaining?) consumer of it. (But in a followup PR after this one)

The current fontconfig needs it too, since it is the mechanism that makes it load the global config.

**Comment by flokli** [comment]
We had enough some staging rounds since https://github.com/NixOS/nixpkgs/pull/93562#pullrequestreview-452120074.

Rebased to latest master.

**Comment by flokli** [comment]
@aszlig You mean, chromium bundles its own version of fontconfig 2.10, which  is still used inside chromium, and it stumbles over our fontconfig version? Wow.

In that case, could we provide a fontconfig 2.10-compatible config (and not build the optional fc caches), so we can keep `fontconfig_210` removed?

**Comment by NickHu** [comment]
NixOS module `fonts.fontconfig.penultimate` still relies on `fontconfig_210` and will trip the assertion. I suppose that the code which generates the fontconfig 2.10 caches in that module can be deleted.

File: nixos/modules/config/fonts/fontconfig.nix [hunk 1]
```diff
@@ -200,59 +176,49 @@ let
...
-    # Latest fontconfig is configured to look for the upstream defaults inside the package.
+    ln -s ${pkg.out}/etc/fonts/conf.d/*.conf \
+          $dst/
+
+    # update 51-local.conf path to look at local.conf
+    rm  $dst/51-local.conf
+
+    substitute ${pkg.out}/etc/fonts/conf.d/51-local.conf \
+               $dst/51-local.conf \
+               --replace local.conf /etc/fonts/${pkg.configVersion}/local.conf
```
Comment (jtojnar) [inline nixos/modules/config/fonts/fontconfig.nix:195]: This is another section that should remain removed from https://github.com/NixOS/nixpkgs/commit/edf2541f02c6b24ea791710d5cadeae36f9b1a3a. See https://github.com/NixOS/nixpkgs/issues/86601#issuecomment-675462227

Comment (flokli) [reply nixos/modules/config/fonts/fontconfig.nix:195]: Not sure I grok this. Can you file a PR to master on what's needed to change?

Comment (jtojnar) [reply nixos/modules/config/fonts/fontconfig.nix:195]: These six lines should have remained removed. I will not get to filing PR before tomorrow.

**Pull request merged**